Libs/MRML/Core/vtkMRMLSliceNode.cxx

```diff
@@ -371,73 +408,194 @@ void vtkMRMLSliceNode::SetOrientationToReformat()
+  this->UpdateMatrices();
 }
 
 //----------------------------------------------------------------------------
 // Local helper to compare matrices -- TODO: is there a standard version of this?
 bool vtkMRMLSliceNode::Matrix4x4AreEqual(const vtkMatrix4x4 *m1, const vtkMatrix4x4 *m2)
 {
-  return vtkAddonMathUtilities::Matrix4x4AreEqual(m1, m2, /* tolerance= */ 0);
+  return vtkAddonMathUtilities::Matrix4x4AreEqual(m1, m2);
+}
+
+//----------------------------------------------------------------------------
+vtkMatrix4x4 *vtkMRMLSliceNode::GetSliceOrientationPreset(const std::string &name)
+{
+  if (!name.compare("Reformat"))
+    {
+    return NULL;
+    }
+
+  std::map<vtkStdString, vtkMatrix4x4*>::iterator it = this->OrientationMatrices.find(name);
+  if (it != this->OrientationMatrices.end())
+    {
+    return it->second;
+    }
+
+  vtkErrorMacro("GetSliceOrientationPreset: invalid Orientation Name: " << name);
+  return NULL;
+}
+
+//----------------------------------------------------------------------------
+std::string vtkMRMLSliceNode::GetSliceOrientationPresetName(vtkMatrix4x4 *sliceToRAS)
+{
+  if (sliceToRAS == NULL)
+    {
+    vtkErrorMacro("GetSliceOrientationPresetName: invalid input Matrix. ");
+    return "";
+    }
+
+  std::map<vtkStdString, vtkMatrix4x4*>::iterator it;
+  for (it = this->OrientationMatrices.begin(); it != this->OrientationMatrices.end(); ++it)
+    {
+    std::string orietation = it->first;
+    if (!orietation.compare("Reformat"))
+      {
+      continue;
+      }
+    vtkMatrix4x4* storedSliceToRAS = this->GetSliceOrientationPreset(orietation);
+    if (storedSliceToRAS != NULL)
+      {
+      if (vtkMRMLSliceNode::Matrix4x4AreEqual(storedSliceToRAS, sliceToRAS))
+        {
+        return orietation;
+        }
+      }
+    }
+
+  vtkErrorMacro("GetSliceOrientationPresetName: orientation string not found.");
+  return "";
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLSliceNode::GetSliceOrientationPresetNames(vtkStringArray *namedOrientations)
+{
+  namedOrientations->SetNumberOfValues(this->OrientationMatrices.size() + 1);
+
+  std::map<vtkStdString, vtkMatrix4x4*>::iterator it;
+  int id = 0;
+  for (it = this->OrientationMatrices.begin(); it != this->OrientationMatrices.end(); ++it)
+    {
+    namedOrientations->SetValue(id, it->first);
+    id++;
+    }
+  namedOrientations->SetValue(id, "Reformat");
+}
+
+//----------------------------------------------------------------------------
+bool vtkMRMLSliceNode::AddSliceOrientationPreset(const std::string &name, vtkMatrix4x4 *sliceToRAS)
+{
+  if (!name.compare("Reformat"))
+    {
+    return false;
+    }
+
+  std::map<vtkStdString, vtkMatrix4x4*>::iterator it = this->OrientationMatrices.find(name);
+  if (it != this->OrientationMatrices.end())
+    {
+    vtkErrorMacro("AddSliceOrientationPreset: the orietation preset " << name << " is already stored.");
+    return false;
+    }
+
```

Comment:
We don't check if the matrix is the same as the matrix in any of the existing presets. This means that we may have multiple names with the same matrix and therefore the result of name lookup based on matrix is undeterministic (especially because we store presets in a map and maps are not sorted).
It would be better to somehow make the behavior more deterministic, for example use a vector instead of a map (and when we search for a preset name, always start from the most recently added preset name).
It's probably useful to be able to have user-defined presets that may have the same orientation as a default preset, but it would be nice if we could ensure that the name lookup would always return the user-defined orientation.